protocols/identify/src/handler.rs

```diff
@@ -346,11 +354,11 @@ impl ConnectionHandler for Handler {
 
         match self.active_streams.poll_unpin(cx) {
             Poll::Ready(Ok(Ok(Success::ReceivedIdentify(remote_info)))) => {
-                self.handle_incoming_info(&remote_info);
-
-                return Poll::Ready(ConnectionHandlerEvent::NotifyBehaviour(Event::Identified(
-                    remote_info,
-                )));
+                if self.handle_incoming_info(&remote_info) {
+                    return Poll::Ready(ConnectionHandlerEvent::NotifyBehaviour(
+                        Event::Identified(remote_info),
+                    ));
+                }
             }
```

Comment:
I think we should do something like `while let Poll::Ready(event) = self.active_stream.poll_unpin(cx) { ...` here, to always drain all events in the stream before returning a `Poll::Pending` below.

Otherwise if the the incoming info is invalid, it won't return the `Event` and instead `Poll::Pending` further below, even though there may still be other events waiting in the active stream.

Admittedly this is an edge case, but I think it's still good to follow the general best practice here.

Reply:
I agree with @elenaf9 that we should drain the stream before we return `Poll::Pending` with the changes done here in this PR in case there are other events waiting to be yielded from stream if `handle_incoming_info` returns `false` (although not related to the PR but maybe we probably should close the connection in this case since `identify` returned a wrong peer id?).

Reply:
> Thanks for taking a look! Agree this issue needs to be handled. Not sure of the best way to drain `self.active_streams` here, I couldn't find a way to iterate it so was trying something like:
>
> ```rust
>  while !self.active_streams.is_empty() {
>	match self.active_streams.poll_unpin(cx) {
>		// ...
>	}
> }
> ```

Why not

```rs
while let Poll::Ready(event) = self.active_streams.poll_unpin(cx) {
    match event {
        Ok(Ok(Success::ReceivedIdentify(remote_info))) => {
            self.handle_incoming_info(&remote_info);

            return Poll::Ready(ConnectionHandlerEvent::NotifyBehaviour(
                Event::Identified(remote_info),
            ));
        }
        ... 
```

I quickly tried it on master branch and all test succeeded.

> (although not related to the PR but maybe we probably should close the connection in this case since identify returned a wrong peer id?).

> Re. closing the connection, sounds like a good idea, I think the only time a remote peer would be sending an invalid pubkey is if they were misconfigured or doing it deliberately.

Agreed. I think it can be included in this PR, but fine with a follow-up PR as well.

Reply:
Ah I was trying something wrong with the `while let` initially and it wouldn't compile, this works!

Updated the PR to use this and opened https://github.com/libp2p/rust-libp2p/issues/5713 for closing the connection